pkg/terraform/state/handler.go

```diff
@@ -76,136 +73,178 @@ func (h *handler) OnChange(key string, obj *v1.State) (*v1.State, error) {
-		runName = name
-		logrus.Debugf("Creating destroy job: %s", runName)
+	logrus.Debug("deploy destroy")
+
+	//no job running, try and run a destroy
+	exec, err := h.deployDestroy(obj, input)
+	if err != nil {
+		logrus.Debug(err)
+		return obj, err
 	}
 
-	execution, err := h.executions.Get(obj.Namespace, runName, metaV1.GetOptions{})
+	obj.Status.ExecutionName = exec.Name
+	v1.StateConditionJobDeployed.True(obj)
+
+	//destroy deployed, setup watcher.
+	execution, err := h.executions.Get(obj.Namespace, obj.Status.ExecutionName, metaV1.GetOptions{})
 	if err != nil {
+		logrus.Debug("error getting the execution")
 		return obj, errors.WithMessage(err, "error getting execution")
 	}
 
-	if v1.ExecutionRunConditionApplied.IsTrue(execution) {
-		err = h.states.Delete(execution.Namespace, execution.Name, &metaV1.DeleteOptions{})
-		if err != nil {
-			if !k8serrors.IsNotFound(err) {
-				return obj, err
-			}
-		}
-		return obj, nil
-	}
+	go h.watchExecution(obj, execution, ActionDestroy)
 
-	if !v1.ExecutionConditionWatchRunning.IsTrue(obj) {
-		go h.watchDestroyRun(obj, execution)
-		v1.ExecutionConditionWatchRunning.True(obj)
+	//return error because no error will clear finalizer even if job did not complete.
+	_, err = h.states.Update(obj)
+	if err != nil {
+		return obj, err
 	}
 
-	return h.states.Update(obj)
+	return obj, fmt.Errorf("watching execution to complete")
 }
 
-// watchDestroyRun checks the execution for the Applied condition, once set
-// terraform destroy was run so the state can be requeued for deletion.
-func (h *handler) watchDestroyRun(state *v1.State, execution *v1.Execution) {
+func (h *handler) watchExecution(state *v1.State, execution *v1.Execution, action string) {
+	logrus.Debugf("About to watch %s", execution.Name)
+	maxIterations := 450 //15min
+	i := 0
 	for {
-		logrus.Debugf("Waiting for %s destroy job.", state.Name)
+		logrus.Debugf("Waiting for execution %s.", execution.Name)
+		if maxIterations < i {
+			return
+		}
+
 		exec, err := h.executions.Get(execution.Namespace, execution.Name, metaV1.GetOptions{})
 		if err != nil {
 			return
 		}
 		if v1.ExecutionRunConditionApplied.IsTrue(exec) {
-			v1.ExecutionConditionWatchRunning.False(state)
-			logrus.Debugf("Destroy complete, requeue %s for deletion.", state.Name)
-			h.states.Enqueue(state.Namespace, state.Name)
+			currentState, err := h.states.Get(state.Namespace, state.Name, metaV1.GetOptions{})
+			if err != nil {
+				logrus.Debug(err)
+				return
+			}
+
+			logrus.Debugf("Execution %s complete", execution.Name)
+			v1.ExecutionConditionWatchRunning.False(currentState)
+			v1.StateConditionJobDeployed.False(currentState)
+			currentState.Status.ExecutionName = ""
+			if ActionDestroy == action {
+				logrus.Debug("Setting Destroyed Condition")
+				v1.StateConditionDestroyed.True(currentState)
+			}
+
+			_, err = h.states.Update(currentState)
+			if err != nil {
+				logrus.Debug("error updating after watching")
```

Comment:
Maybe this shouldn't be a debug? also, use `Debugf` instead of two logs

Reply:
agreed, updated to Errorf